Cache computed properties for toric varieties

HereAround opened the pull request:
- Add two dicts `properties` and `attributes` to struct `NormalToricVarieties`, `AffineNormalToricVarieties`, `ToricDivisors`.
- Use these dicts to cache computed properties.
- Before computation, check if the result is already known/set in these dicts.
- Initialize properties for famous toric varieties (projective space, hirzebruch surface and del Pezzo) upon construction.
- Improve show methods.
- A couple of minor adjustments.

Please note that this PR contains the changes of https://github.com/oscar-system/Oscar.jl/pull/838. In this sense it is a draft and must be rebased once this other PR has been merged.

HereAround committed:
Update tests (otherwise, they test if caching worked upon construction)

HereAround committed:
Remove obsolete whitespaces

HereAround committed:
Remove obsolete code

HereAround committed:
Improve method isprime_divisor

thofma commented:
There is already a mechanism for this using `@declare_other` and `set_special/get_special`, see for example https://github.com/oscar-system/Oscar.jl/blob/79d65df00327fb6009e9c137bbf2bd83e39475b0/src/Rings/FinField.jl#L26-L33
For this the structure needs just one additonal declaration, e.g.
https://github.com/oscar-system/Oscar.jl/blob/d34d698db39583b5fd24ca4e3e375a17684fbdd9/src/Groups/types.jl#L137-L140
@fieker rolled this out, so maybe he has an idea on whether we should try to use this everywhere in Oscar.

fieker commented:
On Mon, Nov 29, 2021 at 02:01:40AM -0800, Tommy Hofmann wrote:
> There is already a mechanism for this using ***@***.***_other` and `set_special/get_special`, see for example https://github.com/oscar-system/Oscar.jl/blob/79d65df00327fb6009e9c137bbf2bd83e39475b0/src/Rings/FinField.jl#L26-L33
> For this the structure needs just one additonal declaration, e.g.
> https://github.com/oscar-system/Oscar.jl/blob/d34d698db39583b5fd24ca4e3e375a17684fbdd9/src/Groups/types.jl#L137-L140
> @fieker rolled this out, so maybe he has an idea on whether we should try to use this everywhere in Oscar.
> 
I thnk it should be used uniformely everywhere - that allows us to
improve it also uniformely everywhere (e.g. rename it...)
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/oscar-system/Oscar.jl/pull/839#issuecomment-981472750

fingolfin commented:
Yeah, we ought to (a) rename `@declare_other` and `set_special/get_special` to something more consistent and sensible, (b) document it and then (c) advertise it to all our devs and beyond

fingolfin commented on src/ToricVarieties/CyclicQuotientSingularities/CyclicQuotientSingularities.jl:
```diff
@@ -6,6 +6,8 @@
 struct CyclicQuotientSingularity <: AbstractNormalToricVariety
     polymakeNTV::Polymake.BigObject
+    properties::Dict{Any,Bool}
+    attributes::Dict{Any,Any}
```
There is really no point in distinguishing attributes and properties here, you can store them all in one dictionary (and then perhaps use the `@declare_other` mechanism).

Also, I would recommend using `Symbol` for the keys. Using methods/functions as index is problematic, and also a lot slower. E.g.:
```julia-repl
julia> D1 = Dict(:dim => 5, :length => 3)
Dict{Symbol, Int64} with 2 entries:
  :dim    => 5
  :length => 3

julia> D2 = Dict(dim => 5, length => 3)
Dict{Function, Int64} with 2 entries:
  dim    => 5
  length => 3

julia> @btime get(D1, :sum, 3);
  21.409 ns (0 allocations: 0 bytes)

julia> @btime get(D1, :dim, 3);
  21.669 ns (0 allocations: 0 bytes)

julia> @btime get(D2, sum, 3);
  258.125 ns (0 allocations: 0 bytes)

julia> @btime get(D2, dim, 3);
  119.801 ns (0 allocations: 0 bytes)
```

fingolfin commented on src/ToricVarieties/LineBundles/constructors.jl:
````diff
@@ -0,0 +1,57 @@
+    ToricLineBundle(v::AbstractNormalToricVariety, c::Vector{fmpz})
+
+Construct the line bundle on the abstract normal toric variety `v` with class `c`.
+```
````
```suggestion
```

fingolfin commented on test/ToricVarieties/runtests.jl:
```diff
@@ -74,7 +74,9 @@ P2 = toric_projective_space(2)
 end
 
-H5 = hirzebruch_surface(5)
+rays = [1 0; 0 1; -1 5; 0 -1]
```
```
ERROR: LoadError: cannot assign a value to variable Oscar.rays from module Main
```

fingolfin commented on test/ToricVarieties/runtests.jl:
```diff
@@ -74,7 +74,9 @@ P2 = toric_projective_space(2)
-H5 = hirzebruch_surface(5)
+rays = [1 0; 0 1; -1 5; 0 -1]
+cones = IncidenceMatrix([[1,2],[2,3],[3,4],[4,1]])
+H5 = NormalToricVariety(PolyhedralFan(rays, cones))
```
```suggestion
rays_ = [1 0; 0 1; -1 5; 0 -1]
cones_ = IncidenceMatrix([[1,2],[2,3],[3,4],[4,1]])
H5 = NormalToricVariety(PolyhedralFan(rays_, cones_))
```

fingolfin commented on test/ToricVarieties/runtests.jl:
```diff
@@ -112,15 +114,25 @@ H5 = hirzebruch_surface(5)
+
+rays = [1 0; 1 1; 0 1; -1 0; -1 -1; 0 -1]
+cones = IncidenceMatrix([[1,2],[2,3],[3,4],[4,5],[5,6],[6,1]])
+dP3 = NormalToricVariety(PolyhedralFan(rays, cones))
```
```suggestion
rays_ = [1 0; 0 1; -1 0; -1 -1]
cones_ = IncidenceMatrix([[1,2],[2,3],[3,4],[4,1]])
dP1 = NormalToricVariety(PolyhedralFan(rays_, cones_))

rays_ = [1 0; 0 1; -1 0; -1 -1; 0 -1]
cones_ = IncidenceMatrix([[1,2],[2,3],[3,4],[4,5],[5,1]])
dP2 = NormalToricVariety(PolyhedralFan(rays_, cones_))

rays_ = [1 0; 1 1; 0 1; -1 0; -1 -1; 0 -1]
cones_ = IncidenceMatrix([[1,2],[2,3],[3,4],[4,5],[5,6],[6,1]])
dP3 = NormalToricVariety(PolyhedralFan(rays_, cones_))
```

fingolfin committed:
Apply suggestions from code review

HereAround committed:
Use @attributes mutable struct for
(*) affine toric varieties
(*) normal toric varieties
(*) cyclic quotient singuarlties

HereAround committed:
Cache properties of normal toric varieties

HereAround committed:
Cache attributes of normal toric varieties

HereAround committed:
Cache betti numbers of normal toric varieties
Change name of method: ith_betti_number -> betti_number

HereAround committed:
Cache attributes of cyclic quotient singularities

HereAround committed:
Cache properties upon construction of normal toric variety

HereAround committed:
Cache properties of toric projective space upon construction

HereAround committed:
Improve computation of class group and Picard group

HereAround committed:
Add method map_from_cartier_divisor_group_to_picard_group
Improve computation of picard group

HereAround committed:
Cache more attributes of toric projective space upon construction

HereAround committed:
Cache properties of hirzebruch surfaces upon construction

HereAround committed:
Cache properties of del Pezzo surfaces upon construction

HereAround committed:
Cache computed quantities for toric divisors

HereAround committed:
Sophisticated show method for normal toric varieties

HereAround committed:
Sophisticated show method for toric divisors

fingolfin commented on src/ToricVarieties/NormalToricVarieties/methods.jl:
```diff
@@ -1,21 +1,40 @@
-    return betti_number
     
+    # if no Betti numbers have been computed
+    betti_numbers = Vector{fmpz}
```
```suggestion
```

HereAround replied:
This line was intended for type stability, as I wanted to save these numbers always as fmpz (rather than Int, Int64). If removed, I might have to change the following line 19 to `betti_numbers = fill(fmpz(-1),2*dim(v)+1)`, i.e. explicitly invoking the cast of `-1` to `fmpz(-1)` there. (Which is of course ok.)

A related question. As there are several Betti numbers for one variety, it seems reasonable to cache them in a vector. Currently, I fill this vector with `-1` at each position. As Betti numbers are always non-negative, a `-1` tells that the corresponding Betti number has not (yet) been computed.

But maybe there is a better way than filling with `-1`s. E.g. to create a vector of suitable length by no entries set and then perform a check if an entry has been set? Suggestions very much appreciated.

This question/issue/challenge repeats for line bundle cohomologies (which I hope to turn to very soon).

fingolfin replied:
This line does not do anything useful: in particular that value is immediately discarded in the following two lines. If you want type stability for this variable, you need to adjust each assignment to it to ensure the right type is assigned, eg via a type assertion and/or an explicit conversion

HereAround replied:
Thank you. Adjusted in https://github.com/oscar-system/Oscar.jl/pull/839/commits/0113ddba20756c92dd5d060a5198f424be4a120e.

fingolfin commented on src/ToricVarieties/CyclicQuotientSingularities/CyclicQuotientSingularities.jl:
```diff
@@ -72,7 +72,12 @@ julia> ecf = cf[1]-1//(cf[2]-fmpq(1,cf[3]))
+    if !has_attribute(cqs, :continued_fraction_hirzebruch_jung)
+        set_attribute!(cqs, :continued_fraction_hirzebruch_jung, Vector{fmpz}(pm_object(cqs).CONTINUED_FRACTION))
+    end
+    return get_attribute(cqs, :continued_fraction_hirzebruch_jung)
```
This and several similar functions could be simplified by using 
```julia
    return get_attribute!(cqs, :continued_fraction_hirzebruch_jung) do
        return Vector{fmpz}(pm_object(cqs).CONTINUED_FRACTION)
    end
```

fingolfin replied:
In general I am thinking about a system to make this stuff even nicer, inspired by GAP but my brain is still mushy

HereAround replied:
Thank you. That is a good suggestion. Indeed, this change can be done in a ton of places. I will do this in a new commit rather than rebasing everything (unless I hear differently).

HereAround replied:
Adjusted in https://github.com/oscar-system/Oscar.jl/pull/839/commits/d9efae27860d632977bdcf636b7a322b4433081f

fingolfin commented on src/ToricVarieties/ToricDivisors/auxilliary.jl:
```diff
@@ -0,0 +1,103 @@
+        end
+    end
+    
+    # torusfactor?
```
?? Doesn't seem to fit the code?

HereAround replied:
Agreed - I will update this comment (which was intended to say "check what we know about the torusfactor").

NVM: This comment is entirely out of place (copied from the corresponding output string from toric varieties)...

I will adjust this.

HereAround replied:
Adjusted in https://github.com/oscar-system/Oscar.jl/pull/839/commits/36f51bff0660d2c7ad0cf17e485b077ceb345483.

fingolfin commented on src/ToricVarieties/ToricDivisors/auxilliary.jl:
```diff
@@ -0,0 +1,103 @@
+    
+    # torusfactor?
+    if last(out_string) == ','
+        out_string = chop(out_string)
```
Personally I'd write this kind of function differently: first create a Vector{String} and push! all those adjectives into it. Then use `join` to concatenate cleverly.

HereAround replied:
Good point. I will adjust accordingly.

HereAround replied:
https://github.com/oscar-system/Oscar.jl/pull/839/commits/36f51bff0660d2c7ad0cf17e485b077ceb345483, https://github.com/oscar-system/Oscar.jl/pull/839/commits/a1368a2c2fe0feb917450b2a5ee91447219af050 have been adjusted accordingly. Let me know what you think.

fingolfin commented on src/ToricVarieties/ToricDivisors/constructors.jl:
```diff
@@ -67,88 +81,9 @@ export DivisorOfCharacter
-###############################################################################
 function Base.show(io::IO, td::ToricDivisor)
-    print(io, "A torus invariant divisor on a normal toric variety")
+    print(io, output_string(td))
```
why this way? Why not instead print into a string when you really need one?

HereAround replied:
My thoughts were no deeper than extending the existing code. As always, I am happy to change to a more robust/preferred/quicker alternative. Are saying that `print(output_string(td))` is better?

fingolfin replied:
No, I am saying that instead  of shoving data into a string and then printing it, it is better to directly print it. And if you sometimes really want a string, you can simply print into a string.

Notw that even `join` takes an io argument and can thus print instead of creating a (temporary) string

HereAround replied:
https://github.com/oscar-system/Oscar.jl/pull/839/commits/36f51bff0660d2c7ad0cf17e485b077ceb345483, https://github.com/oscar-system/Oscar.jl/pull/839/commits/a1368a2c2fe0feb917450b2a5ee91447219af050 have been adjusted accordingly. Let me know what you think.

HereAround committed:
Simplify code for saving/accessing cached information on toric varieties

fingolfin commented on src/ToricVarieties/NormalToricVarieties/attributes.jl:
```diff
@@ -9,7 +9,9 @@
 """
 function dim(v::AbstractNormalToricVariety)
-    return pm_object(v).FAN_DIM::Int
+    return get_attribute!(v, :dim) do
+        return pm_object(v).FAN_DIM::Int
+    end
```
I think some care is needed to retain type stability; the type assertion needs to be moved to the outer return. I think this might work (but didn't test):
```suggestion
    return get_attribute!(v, :dim)::Int do
        return pm_object(v).FAN_DIM
    end
```
You can verify if it worked by (a) calling it (to makes sure I got the syntax right, I am not sure), and (b) using `@code_warntype` on it to verify the return type is inferred correctly.

HereAround replied:
Thank you for the suggestion @fingolfin . For me, this change leads to the following:

> ERROR: LoadError: LoadError: LoadError: syntax: "function" at Oscar.jl/src/ToricVarieties/NormalToricVarieties/attributes.jl:11 expected "end", got "do"

HereAround replied:
For all it matters, here is a MWE via `@code_warntype`, which supports your suspicion that type stability requires more care:

> v = toric_projective_space(2)
A normal, non-affine, smooth, projective, gorenstein, q-gorenstein, fano, 2-dimensional toric variety without torusfactor

> julia> @code_warntype dim(v)
Variables
  #self#::Core.Const(AbstractAlgebra.Generic.dim)
  v::NormalToricVariety
  #1418::Oscar.var"#1418#1419"{NormalToricVariety}

Body::Any
1 ─ %1 = Oscar.:(var"#1418#1419")::Core.Const(Oscar.var"#1418#1419")
│   %2 = Core.typeof(v)::Core.Const(NormalToricVariety)
│   %3 = Core.apply_type(%1, %2)::Core.Const(Oscar.var"#1418#1419"{NormalToricVariety})
│        (#1418 = %new(%3, v))
│   %5 = #1418::Oscar.var"#1418#1419"{NormalToricVariety}
│   %6 = Oscar.get_attribute!(%5, v, :dim)::Any
└──      return %6

Hence, if I read this information correctly, `Body::Any` show that the return type is `Any` rather than `Int`.

HereAround replied:
How about turning function `dim(v::AbstractNormalToricVariety)` into `dim(v::AbstractNormalToricVariety)::Int`? This would seem to work and - based on @code_warntype - enusure that the return value is always `Int64`.

HereAround replied:
I am uncertain if the failures in the runs of the nightly build (ERROR: LoadError: Failed to precompile CxxWrap) are correlated to the changes in this PR.

fingolfin replied:
writing `dim(v::...)::Int` has different meaning, though: it *converts* the result to `Int`, by inserting a call to `convert(Int, ...)`; while a type assertion does just that, it checks that the value is an `Int` (if not, an exception is thrown).

Anyway, let's not worry about this for the moment. I describe a potential future solution for this in another comment.

fingolfin commented on src/ToricVarieties/NormalToricVarieties/attributes.jl:
```diff
@@ -38,8 +40,10 @@ export dim_of_torusfactor
 """
 function euler_characteristic(v::AbstractNormalToricVariety)
-    f_vector = Vector{Int}(pm_object(v).F_VECTOR)
-    return f_vector[dim(v)]
+    return get_attribute!(v, :euler_characteristic) do
+        f_vector = Vector{Int}(pm_object(v).F_VECTOR)
+        return f_vector[dim(v)]
+    end
 end
```
Clearly this is a recurring pattern, and one that I'd expect to see more in the future. It would be good to get this closer to what e.g. GAP (and probably also Polymake and other systems) can do in terms of ease of use and capabilities. E.g. having a general way to test if a value is already known. In GAP parlance, for every attribute `Attr` there is a "tests" `HasAttr` to detect if the attribute value is known. We could do the same and have methods like `has_euler_characteristic` (or `haseuler_characteristic` but I like it more with the underscore?); or perhaps it should be `knows_euler_characteristic`?

Anyway, a potential macro syntax
```julia
@attribute function euler_characteristic(v::AbstractNormalToricVariety)
    f_vector = Vector{Int}(pm_object(v).F_VECTOR)
    return f_vector[dim(v)]
end
```
So really the old code just with `@attribute` prefixed. Also available for inline functions, and it should be engineered to also transfer information about the return type, whether from type assertions or by regular type inference (by using `Base.return_types`)

HereAround replied:
Yes that would be great.

HereAround committed:
Fix bug "cannot assign value to Oscar variable rays"
Rename "rays" to "fan_rays"

fingolfin commented on src/ToricVarieties/NormalToricVarieties/methods.jl:
```diff
@@ -1,21 +1,39 @@
+    if !has_attribute(v, :betti_number)
+        betti_numbers = fill(fmpz(-1),2*dim(v)+1)
+    else
+        betti_numbers = get_attribute(v, :betti_number)
```
To improve type stability, you could have to rewrite this as
```suggestion
        betti_numbers = get_attribute(v, :betti_number)::Vector{fmpz}
```
Alas, I'd actually instead rewrite this function to use `get_attribute!`, and I'd call the attribute `:betti_numbers` (so plural).

HereAround replied:
Sure, let me make these changes.

fingolfin commented on src/ToricVarieties/ToricDivisors/attributes.jl:
```diff
@@ -66,6 +67,19 @@ julia> coefficients(D)
+Return the toric variety of a torus-invariant Weil divisor.
+"""
+function toricvariety(td::ToricDivisor)
+    return get_attribute(td, :toricvariety)
```
This will return `nothing` if `toricvariety` is not set. Is this intentional? Or can it never happen? But if it can never happen, then why is `toricvariety` even an attribute instead of a regular struct field?

If it really should stay an attribute, at least add a type assertion?
```suggestion
    return get_attribute(td, :toricvariety)::AbstractNormalToricVariety
```

HereAround replied:
You are right - let me turn this into a struct field (it will be set upon construction of every toric divisor `td`).

fingolfin reviewed (approved):
Looks good to me.

From my POV we could merge now; the things I mention are truly minor. I'll do it tomorrow or the day after if I hear no complaints

HereAround commented:
> Looks good to me.
> 
> From my POV we could merge now; the things I mention are truly minor. I'll do it tomorrow or the day after if I hear no complaints

Thank you @fingolfin.

HereAround committed:
Improve method betti_number for normal toric variety

Co-authored-by: Max Horn <max@quendi.de>
